### PR TITLE
Updated Lockfile to allow for parsing of new yarn version lock files

### LINKED ISSUE #8054: Yarn 3 compatibility
## Context

One of the takeaways from the Nov 4th, Backstage Openmic event was slow build times increasing becoming an issue for developers using Backstage. I want to start looking at how we can improve this in our project because we find hot module reloading to be fragile and the [existing build system doesn't support extension points that we need](https://github.com/backstage/backstage/issues/7102).

We're going to look at a few different options including [NX Workspaces](http://nx.dev). In the mean time, I wanted to learn about the status of Yarn 3 compatibility.  

## Feature Suggestion

It would be desirable to get bug fixes and performance improvements that come with Yarn 3. Some bug fixes are not getting backported to Yarn 1. Is backstage compatible with yarn 3?

## Possible Implementation

- Unknown

### PR DESCRIPTION
**Pull request by webark**
## Hey, I just made a Pull Request!

Updated the Lockfile to allow for parsing of new versions of lock files that are generated with later yarn versions.
Yarn v1 lock files are still supported.

This does not migrate backstage proper to using yarn 3, but should allow for consumers to choose to use later yarn versions.

This should fix #8054 and help facilitate #9619
#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/cli/src/lib/versioning/Lockfile.ts [hunk 1]
```diff
@@ -79,7 +108,7 @@ export class Lockfile {
 
     for (const [key, value] of Object.entries(data)) {
       const [, name, range] = ENTRY_PATTERN.exec(key) ?? [];
-      if (!name) {
+      if (!name && key !== '__metadata') {
```
Comment (webark) [inline packages/cli/src/lib/versioning/Lockfile.ts:111]: Not sure if this is the he best way, or if there are more special keys. Will look into this a little more.

File: packages/cli/src/lib/versioning/Lockfile.ts [hunk 2]
```diff
@@ -120,7 +150,7 @@ export class Lockfile {
     };
 
     for (const [name, allEntries] of this.packages) {
-      if (filter && !filter(name)) {
+      if (typeof name !== 'string' || (filter && !filter(name))) {
```
Comment (webark) [inline packages/cli/src/lib/versioning/Lockfile.ts:153]: This is an outcome of the previous "__metadata" special key. Probably should be handed up there perhaps.

Comment (webark) [reply packages/cli/src/lib/versioning/Lockfile.ts]: now skipping if key initially if is one of the "special" yarn keys.

**Review by jhaals — APPROVED** [review]
Thanks a lot for driving this forward! 👍
 
I did some manual verification in the form of:

```
# checkout demo repository
yarn set version berry
yarn install
# 6 min of migration and installation
yarn backstage-cli info
# (fails as expected
yarn backstage-cli version:bump
# fails as expected
../backstage/packages/cli/bin/backstage-cli info
# works with this PR
../backstage/packages/cli/bin/backstage-cli versions:bump
# works with pr
git reset --hard origin/master && git clean -fd (back to yarn 1)
../backstage/packages/cli/bin/backstage-cli info
# works
../backstage/packages/cli/bin/backstage-cli versions:bump
# works
```

Perhaps some other of @backstage/maintainers are interested in reviewing but works as I would expect for the things above, not sure if something else should be covered.

**Review by freben — COMMENTED** [review]
LGTM too. @Rugvip ?

**Review by Rugvip — APPROVED** [review]
Nice, looks good, thank you! :grin:

I'm thinking we should be a bit careful and ship this in `1.3-next.0` rather than the mainline release next week? There's a bit further to go to fully roll out Yarn 3 support right? one main thing being v1 -> v3 migration, and maybe back.

**Comment by Rugvip** [comment]
@webark Yep, package management interactions should totally be refactored out and we should support any kind of package manager in the long run. There are some other intricacies like the `prepack` script and stuff like that we need to take care of as well, but I'm sure we can figure that out. Great that you want to help get these things in shape! ❤️ 

By migration I mean some simple `backstage-cli migrate yarn-1-to-3` or something of that sort, which just switches the lockfile format. If that exists in the open I'd rather go with that though, and it could also be shipped as an external tool. The main thing is to not lose the lockfile in the migration to yarn 3, because if that breaks the project it can be a royal pita. Or perhaps that's all unnecessary and it's built into yarn 3? Iirc that wasn't the case when I tried Yarn 2 a while ago though

**Comment by webark** [comment]
> The main thing is to not lose the lockfile in the migration to yarn 3, because if that breaks the project it can be a royal pita. Or perhaps that's all unnecessary and it's built into yarn 3? Iirc that wasn't the case when I tried Yarn 2 a while ago though

I believe this is the case. I tried doing a search for it and didn't find a definite answer though. Haven't tried a practical example though before.

**Comment by jhaals** [comment]
Thanks!

**Pull request merged**